pkg/noderesourcetopology/score.go

```diff
@@ -0,0 +1,117 @@
+package noderesourcetopology
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	apiconfig "sigs.k8s.io/scheduler-plugins/pkg/apis/config"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+)
+
+const (
+	defaultWeight = int64(1)
+)
+
+type scoreStrategy func(v1.ResourceList, v1.ResourceList, resourceToWeightMap) int64
+
+// resourceToWeightMap contains resource name and weight.
+type resourceToWeightMap map[v1.ResourceName]int64
+
+// weight return the weight of the resource and defaultWeight if weight not specified
+func (rw *resourceToWeightMap) weight(r v1.ResourceName) int64 {
+	w, ok := (*rw)[r]
+	if !ok {
+		return defaultWeight
+	}
+
+	if w < 1 {
+		return defaultWeight
+	}
+
+	return w
+}
+
+func (tm *TopologyMatch) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	// calculates the fraction of requested to capacity per each numa-node.
+	// return the numa-node with the minimal score as the node's total score
+	return tm.score(pod, nodeName)
+}
+
+func (tm *TopologyMatch) ScoreExtensions() framework.ScoreExtensions {
+	return nil
+}
+
+// score will use `scoreStrategy` function to calculate the score.
+func (tm *TopologyMatch) score(pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	klog.V(5).Infof("Call score for node %v", nodeName)
+	nodeTopology := findNodeTopology(nodeName, &tm.nodeResTopologyPlugin)
+
+	if nodeTopology == nil {
+		return 0, nil
+	}
+
+	klog.V(5).Infof("nodeTopology: %v", nodeTopology)
+	calculateResourceAllocatableRequest(pod, nodeTopology)
+	resources, allocatablePerNUMA := calculateResourceAllocatableRequest(pod, nodeTopology)
+	return scoreForEachNUMANode(resources, allocatablePerNUMA, tm.scoreStrategy, tm.resourceToWeightMap)
+}
+
+// scoreForEachNUMANode will iterate over all NUMA zones of the node and invoke the scoreStrategy func for every zone.
+// it will return the minimal score of all the calculated NUMA's score, in order to avoid edge cases.
+func scoreForEachNUMANode(requested v1.ResourceList, numaList NUMANodeList, score scoreStrategy, resourceToWeightMap resourceToWeightMap) (int64, *framework.Status) {
+	numaScores := make([]int64, len(numaList))
+	for _, numa := range numaList {
+		numaScores[numa.NUMAID] = score(requested, numa.Resources, resourceToWeightMap)
+	}
+	klog.V(5).Infof("numaScores: %v", numaScores)
+	minScore := findMinScore(numaScores)
+	klog.V(5).Infof("node score: %v", minScore)
+	return minScore, nil
+}
+
+func calculateResourceAllocatableRequest(pod *v1.Pod, nodeTopology *topologyv1alpha1.NodeResourceTopology) (v1.ResourceList, NUMANodeList) {
+	containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
```

Comment:
I recalled I asked the same question to @AlexeyPerevalov before: is it intended to be semantically different from other res calculations? https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/pkg/scheduler/framework/plugins/noderesources/fit.go#L162

Reply:
yes, we followed the way TopologyManager is working
but here in score.go I see only PodLevel behavior of TM, but don't see ContainerLevel.

Reply:
and we don't considering pod.Spec.Overhead, but we should.

Reply:
> I recalled I asked the same question to @AlexeyPerevalov before: is it intended to be semantically different from other res calculations? https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/pkg/scheduler/framework/plugins/noderesources/fit.go#L162

@Huang-Wei I'm not sure it reasonable to generalize both topology and none topology resource calculation, since topology is a hierarchical, currently we use only NUMA, but in future it would be hyper threaded cores and LLC groups of CPU.

Reply:
As long as it's adherent with TM, I'm fine. (it's just every time I look at similar code, I cannot help getting confused...)

Would you mind commenting here, and if possible, paste the link to how TopologyManager calculates the res quest? That would de-confuse other readers as well.

Reply:
@Tal-or could you put the above comment into the code?

Reply:
Sure will add that